backend/danswer/llm/answering/stream_processing/citation_processing.py

```diff
@@ -26,91 +26,161 @@ def extract_citations_from_stream(
+
+    This function effectively processes a stream of text, identifies and reformats citations,
+    and provides both the processed text and citation information as output.
+    """
+
     llm_out = ""
     max_citation_num = len(context_docs)
     curr_segment = ""
-    prepend_bracket = False
     cited_inds = set()
     hold = ""
+
+    raw_out = ""
+    current_citations = []
+    past_cite_count = 0
     for raw_token in tokens:
+        raw_out += raw_token
         if stop_stream:
             next_hold = hold + raw_token
-
             if stop_stream in next_hold:
                 break
-
             if next_hold == stop_stream[: len(next_hold)]:
                 hold = next_hold
                 continue
-
             token = next_hold
             hold = ""
         else:
             token = raw_token
 
-        # Special case of [1][ where ][ is a single token
-        # This is where the model attempts to do consecutive citations like [1][2]
-        if prepend_bracket:
-            curr_segment += "[" + curr_segment
-            prepend_bracket = False
-
         curr_segment += token
         llm_out += token
 
-        possible_citation_pattern = r"(\[\d*$)"  # [1, [, etc
-        possible_citation_found = re.search(possible_citation_pattern, curr_segment)
-
-        citation_pattern = r"\[(\d+)\]"  # [1], [2] etc
-        citation_found = re.search(citation_pattern, curr_segment)
-
-        if citation_found and not in_code_block(llm_out):
-            numerical_value = int(citation_found.group(1))
-            if 1 <= numerical_value <= max_citation_num:
-                context_llm_doc = context_docs[
-                    numerical_value - 1
-                ]  # remove 1 index offset
-
-                link = context_llm_doc.link
-                target_citation_num = doc_id_to_rank_map[context_llm_doc.document_id]
-
-                # Use the citation number for the document's rank in
-                # the search (or selected docs) results
-                curr_segment = re.sub(
-                    rf"\[{numerical_value}\]", f"[{target_citation_num}]", curr_segment
-                )
-
-                if target_citation_num not in cited_inds:
-                    cited_inds.add(target_citation_num)
-                    yield CitationInfo(
-                        citation_num=target_citation_num,
-                        document_id=context_llm_doc.document_id,
+        citation_pattern = r"\[(\d+)\]"
+        print(f"FILTERING FOR CITATIONS {curr_segment}")
+        citations_found = list(re.finditer(citation_pattern, curr_segment))
+
+        if len(citations_found) == 0 and len(llm_out) - past_cite_count > 5:
+            current_citations = []
+
+        if citations_found and not in_code_block(llm_out):
+            # print(f"BEFORE: curr_segment: |{curr_segment}|")
+            last_citation_end = 0
+            length_to_add = 0
+
+            print("CITATIONS")
```

Comment:
Just general advice, since you're using Cursor and you have all of the deployment stuff set up, you should use breakpoints in the debugger instead of adding prints to the code